rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+implementation plan. This parameter should also reflect the state of the
+`mature` parameter: when `mature` is `True`, the new parameter should also be
+`True`. This prepares us to deprecate the `mature` parameter when we remove the
+`unstable` designation from the new parameter.
+
+When `unstable__include_sensitive_results` is `False` (default behaviour), query
+the filtered index by appending the `-filtered` suffix to the index name. Update
+the existing `mature` parameter handling to check
+`unstable__include_sensitive_results`. When the new parameter is `False`,
+results with the `mature` set to `true` should also be filtered out.
+
+When `unstable__include_sensitive_results` is `True`, query the origin index
+(existing behaviour).
+
+### Derive the list of sensitive document IDs based on the filtered index (overview step 4)
+
+When `unstable__include_sensitive_results` is `True`, we need to derive which
+documents in the results have sensitive textual content. To do this, we will
+rely on the fact that those documents are _not_ in the filtered index.
+Therefore, we can pull the list of `_id`s for the documents retrieved from the
+origin index and query for their presence in the filtered index. If they are in
+the filtered index, then we know they do not have sensitive textual content.
+Something along the following lines of pseudocode:
+
+```py
+results = query_origin_index()
+result_ids = set({r["_id"] for r in results})
+
+results_in_filtered_index = query_filtered_index(results_ids)
+ids_in_filtered_index = set({r["_id"] for r in results_in_filtered_index})
+
+# Use set arithmatic to derive the list of sensitive documents
+sensitive_text_result_ids = result_ids - ids_in_filtered_index
+```
+
+Add the resulting `sensitive_text_result_ids` set to the result serialiser
+context so that the media serialisers can reference them to derive the sensitive
+field:
+
+```py
+sensitivity = serializers.SerializerMethodField()
+
+def get_sensitivity(self, obj):
+    result = []
+    if obj["mature"]:
+        result.append("provider_supplied_mature")
+
+    if obj["identifier"] in self.context["sensitive_result_ids"]:
+        result.append("sensitive_text")
+
+    return result
+```
+
+#### Mature results
+
+Results with the `mature` field set to `true` will still be included in the
+"filtered" index. While it is easy to exclude these results from the filtered
+index, we cannot do so without making it harder to derive the list of results
+with sensitive textual content based on presence in the filtered index. This is
+because if the filtered index also excludes documents marked mature, then the
+`origin_result_ids - filtered_result_ids` will generate a set exclusive of
+documents marked mature _and_ those with sensitive textual content. Critically
+this means that results in the origin index that are mature but do not have
+sensitive text would get marked as having sensitive text based on their presence
+in the `sensitve_result_ids` set.
```

Comment:
Just double checking my understanding:  the goal is to make it possible to determine if an API result was marked sensitive due to `provider_supplied_mature` vs matching `sensitive_text`, but (at least for now) they will be treated exactly the same.

Reply:
Ah, that makes a lot of sense! Thank you for that link.